packages/drivers/odsp-driver/src/getFileLink.ts

```diff
@@ -90,6 +93,86 @@ export async function getFileLink(
 	return fileLink;
 }
 
+async function getRequestInformation(
+	getToken: TokenFetcher<OdspResourceTokenFetchOptions>,
+	odspUrlParts: IOdspUrlParts,
+	logger: ITelemetryLoggerExt,
+	options: TokenFetchOptionsEx,
+	fileItem: FileItemLite,
+) {
+	const storageTokenFetcher = toInstrumentedOdspTokenFetcher(
+		logger,
+		odspUrlParts,
+		getToken,
+		true /* throwOnNullToken */,
+	);
+	const storageToken = await storageTokenFetcher(options, "GetFileLinkCore");
+	assert(
+		storageToken !== null,
+		0x2bb /* "Instrumented token fetcher with throwOnNullToken = true should never return null" */,
+	);
+
+	// IMPORTANT: In past we were using GetFileByUrl() API to get to the list item that was corresponding
+	// to the file. This was intentionally replaced with GetFileById() to solve the following issue:
+	// GetFileByUrl() uses webDavUrl to locate list item. This API does not work for Consumer scenarios
+	// where webDavUrl is constructed using legacy ODC format for backward compatibility reasons.
+	// GetFileByUrl() does not understand that format and thus fails. GetFileById() relies on file item
+	// unique guid (sharepointIds.listItemUniqueId) and it works uniformly across Consumer and Commercial.
+	const { url, headers } = getUrlAndHeadersWithAuth(
+		`${
+			odspUrlParts.siteUrl
+		}/_api/web/GetFileById(@a1)/ListItemAllFields/GetSharingInformation?@a1=guid${encodeURIComponent(
+			`'${fileItem.sharepointIds.listItemUniqueId}'`,
+		)}`,
+		storageToken,
+		true,
+	);
+	const requestInit = {
+		method: "POST",
+		headers: {
+			"Content-Type": "application/json;odata=verbose",
+			"Accept": "application/json;odata=verbose",
+			...headers,
+		},
+	};
+
+	return { url, requestInit };
+}
+
+async function fetchWithLocationRedirectionHandling<T>(
+	getToken: TokenFetcher<OdspResourceTokenFetchOptions>,
+	odspUrlParts: IOdspUrlParts,
+	logger: ITelemetryLoggerExt,
+	options: TokenFetchOptionsEx,
+	fileItem: FileItemLite,
+): Promise<IOdspResponse<Response>> {
+	for (;;) {
+		const { url, requestInit } = await getRequestInformation(
+			getToken,
+			odspUrlParts,
+			logger,
+			options,
+			fileItem,
+		);
+
+		try {
+			return await fetchHelper(url, requestInit);
+		} catch (error: any) {
+			if (
+				error.errorType !== DriverErrorTypes.fileNotFoundOrAccessDeniedError ||
+				!error.redirectLocation
+			) {
+				throw error;
+			}
+
+			const newSiteDomain = new URL(error.redirectLocation).origin;
```

Comment:
Is this right? We can get legit 404 because there is no such file at all (no redirection is involved).
We will get here, likely with undefined error.redirectLocation and will crash, no?

Reply:
I think this is right, no? On line 163 I am checking that this was the error that I threw, and that redirect Location is set before retrying. If neither of those are true, I'm just re-throwing the error

Reply:
Yes, looks good!